### DISCUSSION TITLE
Proper use of BeginAbsoluteSequence in UpdateStateTransforms versus ApplyCustomUpdateState?

### DISCUSSION THREAD

**Discussion by goodtrailer** [ppy/osu#18996] [Q&A] [answered]
It seems there's been a bit of a difference in how animations have been applied when overriding `DrawableHitObject.UpdateStateTransforms` versus when subscribing to `DrawableHitObject.ApplyCustomUpdateState`. When overriding `UpdateStateTransforms`, the convention has been to apply animations "directly", such as here: https://github.com/ppy/osu/blob/50a546244668e46d0be75bb2e548facecb45dbba/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSlider.cs#L316-L335 ([additional example](https://github.com/ppy/osu/blob/50a546244668e46d0be75bb2e548facecb45dbba/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs#L187-L210))

---

But when subscribing to `ApplyCustomUpdateState`, the convention has been to use a `BeginAbsoluteSequence`, such as here: https://github.com/ppy/osu/blob/50a546244668e46d0be75bb2e548facecb45dbba/osu.Game.Rulesets.Osu/Skinning/Default/MainCirclePiece.cs#L77-L110 ([additional example](https://github.com/ppy/osu/blob/4bc26dbb487241e2bbae73751dbe9e93a4e427da/osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyMainCirclePiece.cs#L128-L161))

---

What's the reason for this? Is there a difference between `UpdateStateTransforms` and `ApplyCustomUpdateState` that I'm not understanding? I can understand using `BeginAbsoluteSequence` for aligning animations to the start time, since there is no `ArmedState` that aligns to the start time. But `ArmedState.Hit` and `ArmedState.Miss` are received at the end time, so it doesn't seem necessary then (and my limited testing seems to support this). If it's just for safety's sake in case of misalignment, why isn't it deemed necessary when overriding `UpdateStateTransforms`?

**Comment by smoogipoo** [comment, marked as answer]
Perhaps I'm misunderstanding the question, but the contents of the `UpdateStateTransforms` methods are already pinned to the state transition times, whereas `ApplyCustomUpdateState` isn't:

https://github.com/ppy/osu/blob/65917fbb99dfa95e7eb0f0f0555513ee5d78ec7b/osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs#L421-L436

If your question was why you need to use `HitStateUpdateTime` instead of just a condition checking the armed state for hit or miss, then I believe the reason is seeking and especially non-framestable (60fps) seeks such as in the editor. If the editor is seeked to +2000ms relative to a hitobject, the hit state transforms should still be at +160ms or whatever the miss window of the hitobject is.

I can't remember how well this works in practice, though (and we can probably do it better by faking a current time in `DrawableHitObject.OnKilled()`).
